drivers: charger: max20335 make cfg complete

LukaszMadejGrinn opened the pull request:
Allow to configure more charger parameters from device tree.
Add few fixes to existing code.

LukaszMadejGrinn committed:
drivers: charger: charger_max20355: fix configuration

Maximum charging current is selected with an external resistor in
the MAX20335 charger. Therefore it is not possible to configure it
with software directly. There is only a capability to limit current
set with hardware but configuration of the limiter is not
straight-forward.

To reflect real functionality, drop usage of
`constant-charge-current-max-microamp` property as an required one and
use custom `chgin-to-sys-current-limit-microamp` instead.
Use enum in binding file to signalize improper values at compile time.

Drop support for `CHARGER_PROP_CONSTANT_CHARGE_CURRENT_UA` API property
since this cannot be handled.

The `max20335_get_constant_charge_current()` function become useless so
remove it.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
dts: bindings: charger: max20335: use enum for property

Add enum for `constant-charge-voltage-max-microvolt` property to
signalize improper values given with device tree at compile time.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: charger: charger_max20335: drop redundant validation

`max20335_get_constant_charge_current()` and
`max20335_set_chgin_to_sys_current_limit()` functions has redundant input
values validation. Drop not important ones to make the code clean.
While on it remove `max20335_get_constant_charge_voltage()` function
since it become useless.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: charger: charger_max20335: fix linear range usage

Fix validation of `linear_range_get_index()` results.
The function can resturn both 0 or -ERANGE but -EINVAL is never
returned. Use comparison against 0 to make the validation robust.

Use valid maximum index for the linear range. The 0x0C is reserved
value and should not be used. Therefore replace it with 0x0B.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: charger: charger_max20335: rename macro to reduce ambiguity

The ILimCntl name can be understood as both register name and its field
name. Therefore it is better to change macro name so it contains both
register and field name. Second field of this register will be also
utilized by this driver in the future and new code will be aligned to
this naming convention.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: charger: charger_max20335: rename register related macros

Rename macros related to charger harware registers to make them
consistent.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: charger: charger_max20335: handle minimum system voltage

Allow to configure minimum system voltage threshold parameter.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: charger: charger_max20355: handle recharge threshold

Allow to configure recharge threshold.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: charger: charger_max20355: handle thermistor mode

Allow to configure thermistor monitoring mode.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

bbilas commented on dts/bindings/charger/maxim,max20335-charger.yaml:
```diff
@@ -8,12 +8,20 @@ include: battery.yaml
   constant-charge-voltage-max-microvolt:
     required: true
 
+  chgin-to-sys-current-limit-microamp:
```
I would set it as a required property or use one of the enum values as the default one as this chip has a couple of variants with different default register values. Then you could drop the `PROP_NOT_USED` from the code.

LukaszMadejGrinn replied:
If you want to drop `PROP_NOT_USED` I would prefer to use all properites as required.
As you said, there is several chip variants which differs in default configuration.
It would be safer to set all configuration entries correctlly based on the battery spec.

LukaszMadejGrinn replied:
Modified in a way written above.

bbilas replied:
LGTM! Please check if we need to add a note in the release-note about that.

bbilas commented on dts/bindings/charger/maxim,max20335-charger.yaml:
```diff
@@ -35,6 +35,22 @@ properties:
       CHGIN to SYS path current limitter configuration.
       Refer to ILimCntl register description for details.
 
+  system-voltage-min-threshold-microvolt:
```
ditto

LukaszMadejGrinn replied:
Done.

bbilas commented on dts/bindings/charger/maxim,max20335-charger.yaml:
```diff
@@ -51,6 +51,17 @@ properties:
       voltage drops to this level, the charger current
       is reduced to prevent battery damage.
 
+  recharge-threshold-microvolt:
```
ditto

LukaszMadejGrinn replied:
Done.

bbilas commented on drivers/charger/charger_max20335.c:
```diff
@@ -343,10 +385,30 @@ static int max20335_init_properties(const struct device *dev)
 	return 0;
 }
 
+enum charger_max20335_therm_mode max20335_string_to_therm_mode(char *mode_string)
```
`const char *mode_string`

LukaszMadejGrinn replied:
Fixed

bbilas commented on dts/bindings/charger/maxim,max20335-charger.yaml:
```diff
@@ -62,6 +62,17 @@ properties:
       Recharge threshold in relation to BatReg.
       Refer to ChgCntlA register description for details.
 
+  thermistor-monitoring-mode:
```
ditto

LukaszMadejGrinn replied:
Done.

rriveramcrus commented on drivers/charger/charger_max20335.c:
```diff
@@ -389,23 +389,8 @@ static int max20335_set_prop(const struct device *dev, charger_prop_t prop,
-			data->charge_voltage_uv = val->const_charge_voltage_uv;
-		}
-
-		return ret;
```
I disagree with this commit, I can think of a few reasons why an application may need to adjust these targets. 

https://home.jeita.or.jp/upload_file/20210407112651_JbDgUSWGiB.pdf
What if the charging task wants to adjust the current (CC) and voltage (CV) targets based on the thermistor (reducing CC or CV if it is too hot or cold to charge at full capacity, safely)?

What if the system knows the user is asleep and wants to charge at a lower CC overnight?

How about if you clamp the maximum settable CC and CV targets to what is described in devicetree and allow runtime configuration instead? Please let me know if this comment is clear enough.

LukaszMadejGrinn replied:
I think I got your point. Now I see two issues with this commit.

1) Order of commits is not correct. This one should be applied after https://github.com/zephyrproject-rtos/zephyr/pull/69910/commits/21537aeeab0a57391366303c16854db7cd19ae4b
2) Based on your input, `CHARGER_PROP_CONSTANT_CHARGE_VOLTAGE_UV` should stay.

If an application needs to set `CHARGER_PROP_CONSTANT_CHARGE_CURRENT_UA`, the property should be added again but different code needs to be called underneath (https://github.com/zephyrproject-rtos/zephyr/pull/69910/commits/21537aeeab0a57391366303c16854db7cd19ae4b again).
I propose to cover this within a separate PR.

Please confirm that we are on the same page.

rriveramcrus replied:
We are on the same page. 
I am okay with leaving reimplementing the `CHARGER_PROP_CONSTANT_CHARGE_CURRENT_UA` setter properly for a future PR. Given that the power stage is linear, the `CHARGER_PROP_CONSTANT_CHARGE_CURRENT_UA` is set by the input current limit indirectly so moving the value around at runtime could introduce new issues.

LukaszMadejGrinn replied:
Hi @rriveramcrus 
I have reworked discussed part. Can you re-review please ?

rriveramcrus commented on dts/bindings/charger/maxim,max20335-charger.yaml:
```diff
@@ -8,11 +8,70 @@ include: battery.yaml
+      voltage drops to this level, the charger current
+      is reduced to prevent battery damage.
+
+  recharge-threshold-microvolt:
```
https://www.kernel.org/doc/Documentation/devicetree/bindings/power/supply/battery.yaml

We may want to have this property name match Linux as `re-charge-voltage-microvolt` and also add it to `zephyr/dts/bindings/battery/battery.yaml`

bbilas replied:
It seems it requires a note in migration-guide.

rriveramcrus replied:
@LukaszMadejGrinn, if we resolve this, I am ready to approve

LukaszMadejGrinn replied:
@rriveramcrus Its done. Please have a look. I have also added some notes to release notes and migration guide but I'm not sure if I've done this correctly.

LukaszMadejGrinn committed:
dts: battery: add property

Add `re-charge-voltage-microvolt` property allowing to set
limit to automatically start charging again.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

bbilas commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -74,6 +74,15 @@ Analog-to-Digital Converter (ADC)
+Charger
+=======
+
+* Dropped ``constant-charge-current-max-microamp`` property in ``charger_max20335`` driver because
```
@kartben Could you please check if we should mention this in the migration guide?
